parser/parser.go

```diff
@@ -158,30 +158,41 @@ func (re *recursionError) Error() string {
 var _ error = &recursionError{}
 
 type recursionListener struct {
-	maxDepth int
-	depth    int
+	maxDepth      int
+	ruleTypeDepth map[int]int
 }
 
 func (rl *recursionListener) VisitTerminal(node antlr.TerminalNode) {}
 
 func (rl *recursionListener) VisitErrorNode(node antlr.ErrorNode) {}
 
 func (rl *recursionListener) EnterEveryRule(ctx antlr.ParserRuleContext) {
-	if ctx != nil && ctx.GetRuleIndex() == gen.CELParserRULE_expr {
-		if rl.depth >= rl.maxDepth {
-			rl.depth++
-			panic(&recursionError{
-				message: fmt.Sprintf("expression recursion limit exceeded: %d", rl.maxDepth),
-			})
-		}
-		rl.depth++
+	if ctx == nil {
+		return
+	}
+	ruleIndex := ctx.GetRuleIndex()
+	depth, found := rl.ruleTypeDepth[ruleIndex]
+	if !found {
+		depth = 1
+	} else {
+		depth++
+	}
+	rl.ruleTypeDepth[ruleIndex] = depth
+	if depth >= rl.maxDepth {
+		panic(&recursionError{
+			message: fmt.Sprintf("expression recursion limit exceeded: %d", rl.maxDepth),
+		})
 	}
 }
 
 func (rl *recursionListener) ExitEveryRule(ctx antlr.ParserRuleContext) {
-	if ctx != nil && ctx.GetRuleIndex() == gen.CELParserRULE_expr {
-		rl.depth--
+	if ctx == nil {
+		return
 	}
+	ruleIndex := ctx.GetRuleIndex()
+	depth := rl.ruleTypeDepth[ruleIndex]
```

Comment:
I added a guard to make sure that the rule counter is found and > 0 before decrementing.